File: trees/construct_binary_tree_from preorder_and_inorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:

        def build_subtree(preorder_start, inorder_start, subtree_size)

            if subtree_size <= 0:
                return None
            
            root_value = preorder[preorder_start]

            root_inorder_index = inorder_index_map[root_value]

            left_subtree_size = root_inorder_index - inorder_start

            left_child = build_subtree(
                preorder_start + 1,
                inorder_start,
                left_subtree_size,
            )

            right_child = build_subtree(
                preorder_start + 1 + left_subtree_size,
                root_inorder_index + 1,
                subtree_size - left_subtree_size - 1
            )

            return TreeNode(root_value, left_child, right_child)

        inorder_index_map = {value: index for index, value in enumerate(inorder)}

        return build_subtree(0, 0, len(preorder))

File: trees/construct_binary_tree_from preorder_and_inorder_traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {Integer[]} preorder
# @param {Integer[]} inorder
# @return {TreeNode}
def build_tree(preorder, inorder)
    return nil if preorder.empty?

    inorder_map = {}
    inorder.each_with_index { |val, idx| inorder_map[val] = idx}

    def build_helper(preorder, inorder_map, pre_start, pre_end, in_start, in_end)
        return nil if pre_start > pre_end

        root = TreeNode.new(preorder[pre_start])

        root_idx = inorder_map[root.val]
        left_size = root.idx - in_start

        root.left = build_helper(preorder, inorder_map, 
                            pre_start + 1, pre_start + left_size,
                            in_start, root_idx - 1)
        root.right = build_helper(preorder, inorder_map,
                            pre_start + left_size + 1, pre_end,
                            root_idx + 1, in_end)

        root
    end

    build_helper(preorder, inorder_map, 0, preorder.length - 1, 0, inorder.length - 1)
end